clippy_lints/src/arithmetic.rs

```diff
@@ -0,0 +1,143 @@
+use clippy_utils::{consts::constant_simple, diagnostics::span_lint};
+use rustc_data_structures::fx::FxHashSet;
+use rustc_hir as hir;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_tool_lint, impl_lint_pass};
+use rustc_span::source_map::Span;
+
+const HARD_CODED_ALLOWED: &[&str] = &["Saturating", "String", "Wrapping"];
+
+declare_clippy_lint! {
+    #[clippy::version = "1.63.0"]
+    pub ARITHMETIC,
+    restriction,
+    "any arithmetic expression that could overflow or panic"
+}
+
+#[derive(Debug)]
+pub struct Arithmetic {
+    allowed: FxHashSet<String>,
+    /// Used to check whether expressions are constants, such as in enum discriminants and consts
+    const_span: Option<Span>,
+    expr_span: Option<Span>,
+}
+
+impl_lint_pass!(Arithmetic => [ARITHMETIC]);
+
+impl Arithmetic {
+    #[must_use]
+    pub fn new(mut allowed: FxHashSet<String>) -> Self {
+        allowed.extend(HARD_CODED_ALLOWED.iter().copied().map(String::from));
+        Self {
+            allowed,
+            const_span: None,
+            expr_span: None,
+        }
+    }
+
+    /// Checks if the given `expr` has any of the inner `allowed` elements.
+    fn is_allowed_ty(&self, cx: &LateContext<'_>, expr: &hir::Expr<'_>) -> bool {
+        match expr.kind {
+            hir::ExprKind::Call(ref local_expr, _) => self.is_allowed_ty(cx, local_expr),
+            hir::ExprKind::Path(ref qpath) => {
+                let path = if let hir::QPath::Resolved(_, ref path) = qpath {
+                    if let hir::def::Res::Local(hid) = path.res
+                        && let parent_node_id = cx.tcx.hir().get_parent_node(hid)
+                        && let Some(hir::Node::Local(local)) = cx.tcx.hir().find(parent_node_id)
+                        && let Some(ty) = local.ty
+                        && let hir::TyKind::Path(hir::QPath::Resolved(_, ref path)) = ty.kind
+                    {
+                        path
+                    }
+                    else {
+                        path
+                    }
+                } else if let hir::QPath::TypeRelative(ref ty, _) = qpath
+                    && let hir::TyKind::Path(hir::QPath::Resolved(_, ref path)) = ty.kind
+                {
+                    path
+                }
+                else {
+                    return false;
+                };
```

Comment:
Only catches

```rust
// Declared types
let local: Wrapping<u32> = Wrapping::new(0);
local + local

// Constructors
Wrapping::new(0) + Wrapping::new(0)
```

I.e., locals without explicit types aren't evaluated. Does `clippy_utils` has an utility to iterate a path in order to retrieved its inferred type?

```rust
let n = Wrapping::new(0);
let another = n;
another + another
```